### PR TITLE
Overhaul AI

### PR DESCRIPTION
**Pull request by ghoulslash**
## Description
- Convert AI battle scripts into C functions
- Expand current AI checks to include previously-neglected effects/abilities
- Add some AI handicap flags/other flags as examples

Thanks to Syreldar for some testing/fixes

### CODE REVIEW

**Comment by luckytyphlosion** [comment]
There's a part of the AI that I noticed is bugged. Over [here](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_ai_main.c#L276-L297), the AI will return with `AI_CHOICE_SWITCH` after some conditions are met. However, following [the](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_controller_opponent.c#L1559-L1570) [trace](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_main.c#L3961-L3979), this doesn't seem to actually do anything. It'll just emit a value of 10 and 0xffff which will just redirect execution back to [STATE_BEFORE_ACTION_CHOSEN](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_main.c#L3737).

What I would do is consolidate the choice of whether to switch or use a move into one single action, like I've done within [this commit](https://github.com/luckytyphlosion/pokeemerald-expansion/commit/6230b738dd922ca12e75987cf3bb40ad17d0a53f). This way, the code for switching doesn't need to stay separate from the code for choosing the move to use.

**Comment by ghoulslash** [comment]
> 
> 
> There's a part of the AI that I noticed is bugged. Over [here](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_ai_main.c#L276-L297), the AI will return with `AI_CHOICE_SWITCH` after some conditions are met. However, following [the](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_controller_opponent.c#L1559-L1570) [trace](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_main.c#L3961-L3979), this doesn't seem to actually do anything. It'll just emit a value of 10 and 0xffff which will just redirect execution back to [STATE_BEFORE_ACTION_CHOSEN](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_main.c#L3737).
> 
> What I would do is consolidate the choice of whether to switch or use a move into one single action, like I've done within [this commit](https://github.com/luckytyphlosion/pokeemerald-expansion/commit/6230b738dd922ca12e75987cf3bb40ad17d0a53f). This way, the code for switching doesn't need to stay separate from the code for choosing the move to use.

I started looking into this, and I'm not convinced the bug exists. If you replace the [highlighted code](https://github.com/ghoulslash/pokeemerald/blob/1ae57f26a37f1469e59ceb920df4a28e40399a29/src/battle_ai_main.c#L276-L297) with a forced switch as follows, the trainer switches every turn as expected.
```
AI_THINKING_STRUCT->switchMon = TRUE;
return AI_CHOICE_SWITCH;
```

**Comment by luckytyphlosion** [comment]
> > There's a part of the AI that I noticed is bugged. Over [here](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_ai_main.c#L276-L297), the AI will return with `AI_CHOICE_SWITCH` after some conditions are met. However, following [the](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_controller_opponent.c#L1559-L1570) [trace](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_main.c#L3961-L3979), this doesn't seem to actually do anything. It'll just emit a value of 10 and 0xffff which will just redirect execution back to [STATE_BEFORE_ACTION_CHOSEN](https://github.com/ghoulslash/pokeemerald/blob/1ae57f2/src/battle_main.c#L3737).
> > What I would do is consolidate the choice of whether to switch or use a move into one single action, like I've done within [this commit](https://github.com/luckytyphlosion/pokeemerald-expansion/commit/6230b738dd922ca12e75987cf3bb40ad17d0a53f). This way, the code for switching doesn't need to stay separate from the code for choosing the move to use.
> 
> I started looking into this, and I'm not convinced the bug exists. If you replace the [highlighted code](https://github.com/ghoulslash/pokeemerald/blob/1ae57f26a37f1469e59ceb920df4a28e40399a29/src/battle_ai_main.c#L276-L297) with a forced switch as follows, the trainer switches every turn as expected.
> 
> ```
> AI_THINKING_STRUCT->switchMon = TRUE;
> return AI_CHOICE_SWITCH;
> ```

Hmm, then I'd need to check the trace within no$gba to ensure that no bad side effects are happening then. In any case, it would still be better to rewrite AI choice so that switching AND move selection are done at the same time (there's no reason why they have to be separate).

File: include/constants/battle.h [hunk 1]
```diff
@@ -220,6 +220,9 @@
 #define SIDE_STATUS_CRAFTY_SHIELD           (1 << 20)
 #define SIDE_STATUS_MAT_BLOCK               (1 << 21)
 
+#define SIDE_STATUS_HAZARDS_ANY    (SIDE_STATUS_SPIKES | SIDE_STATUS_STICKY_WEB | SIDE_STATUS_TOXIC_SPIKES | SIDE_STATUS_STEALTH_ROCK)
+#define SIDE_STATUS_SCREEEN_ANY    (SIDE_STATUS_REFLECT | SIDE_STATUS_LIGHTSCREEN | SIDE_STATUS_AURORA_VEIL)
```
Comment (MandL27) [inline include/constants/battle.h:224]: Should probably be `SIDE_STATUS_SCREEN_ANY` (note the number of Es)

**Review by MandL27 — APPROVED** [review]
Lgtm. Not merging yet in case you want to fix the typo I spotted.

**Review by MandL27 — APPROVED** [review]

**Pull request merged**